test/extended/testdata/gssapi/fedora/base/gssapi-tests.sh

```diff
@@ -0,0 +1,280 @@
+
+CLIENT_HAS_GSSAPI=true
+SERVER_HAS_BASIC=true
+users=(user1 user2 user3 user4 user5)
+
+# Client has no GSSAPI and server is GSSAPI only
+# Everything fails
+
+if [[ -z CLIENT_HAS_GSSAPI && -z SERVER_HAS_BASIC ]]; then
+    for u in "${users[@]}"; do
+        full="$u$realm"
+        os::cmd::expect_failure_and_text 'oc login' 'No Kerberos credentials available'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text 'oc login' 'No Kerberos credentials available'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text "oc login -u $full -p wrongpassword" "Can't find client principal $full in cache collection"
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text "oc login -u $full -p password" "Can't find client principal $full in cache collection"
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text "oc login -u $u -p wrongpassword" 'An invalid name was supplied'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text "oc login -u $u -p password" 'An invalid name was supplied'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+    done
+fi
+
+# Client has no GSSAPI and server is GSSAPI with Basic fallback
+# Only BASIC works
+
+if [[ -z CLIENT_HAS_GSSAPI && -n SERVER_HAS_BASIC ]]; then
+    for u in "${users[@]}"; do
+        full="$u$realm"
+        os::cmd::expect_failure_and_text 'oc login' 'Login failed (401 Unauthorized)'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text 'oc login' 'Login failed (401 Unauthorized)'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text "oc login -u $full -p wrongpassword" 'Login failed (401 Unauthorized)'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_failure_and_text "oc login -u $u -p wrongpassword" 'Login failed (401 Unauthorized)'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+
+        os::cmd::expect_success_and_text "oc login -u $full -p password" 'Login successful.'
+        os::cmd::expect_success_and_text 'oc whoami' $u
+        os::cmd::expect_success_and_text 'oc logout' $u
+
+        os::cmd::expect_success_and_text "oc login -u $u -p password" 'Login failed (401 Unauthorized)'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+    done
+fi
+
+# Client has GSSAPI and server is GSSAPI only
+# Only GSSAPI works
+
+if [[ -n CLIENT_HAS_GSSAPI && -z SERVER_HAS_BASIC ]]; then
+    for u in "${users[@]}"; do
+        os::cmd::expect_failure "kinit $u <<< wrongpassword"
+        os::cmd::expect_failure_and_text 'oc login' 'No Kerberos credentials available'
+        os::cmd::expect_failure_and_not_text 'oc whoami' $u
+        os::cmd::expect_success 'kdestroy'
+
```

Comment:
test combinations of kinit with unqualified and realm-qualified names, and oc login without username, with short username, and with realm-qualified username, and make sure whoami always returns the fully-qualified username